Add `NonEmptyList/fromList`

joneshf opened the pull request:
As discussed on slack, we can make this function so it's a bit easier to do other things with `NonEmptyList`.

Let me know if you'd like anything done differently.

joneshf committed:
Add `NonEmptyList/fromList`

We can provide an easy way to convert back to a list.

We expect that `NonEmptyList/fromList` and `NonEmptyList/toList`
form a codec pair.
I.e.:
* `NonEmptyList/fromList (NonEmptyList/toList x) = Some x`
* `Optional/map NonEmptyList/toList (NonEmptyList/fromList x) = Some x`

FintanH commented on NonEmptyList/fromList:
```diff
@@ -0,0 +1,16 @@
+→ List/fold
+  a
+  list
+  (Optional (./Type a))
```
One minor nit: Since we use `./Type` in multiple places could we use a let binding:
```haskell
let NonEmptyList = ./Type
in ...
```

The reason for this is if we start to use freezing then we won't have multiple SHA diffs.

joneshf replied:
I followed the convention of naming it `NonEmpty`. I hope that's okay. Happy to change it to `NonEmptyList`, if not.

FintanH replied:
I think `NonEmptyList` would be better since that's the convention we've gone for in the directory naming :)

FintanH commented on NonEmptyList/fromList:
```diff
@@ -0,0 +1,16 @@
+  λ(a : Type)
```
Can we add the following documentation, feel free to reword or change.


{-
Create a NonEmptyList from a List as long as there is at least one element in the List. Otherwise return None.

Examples:

```
./NonEmptyList/fromList Natural [1, 2, 3, 4, 5]
= Some { head = 1, tail = [ 2, 3, 4, 5 ] }

./NonEmptyList/fromList Natural ([] : List Natural)
= None { head : Natural, tail : List Natural }
```
-}

FintanH reviewed (changes requested):
This is awesome! Just some tidying up and then we can merge :ok_hand: :100:

joneshf committed:
Extract `./Type` import

If we start freezing imports,
having the same import in multiple places will lead to noisy diffs.
To nip this in the bud,
we import once and use the binding multiple times.

We follow the convention of naming the type `NonEmpty`.

joneshf committed:
Document `NonEmptyList/fromList`

Providing some human readable description of what expressions do,
along with examples of usage, should help others understand it better.

joneshf committed:
Rename the type

We name to follow the directory structure
rather than what other files are doing.

FintanH commented:
Thanks @joneshf :tada: :heart: Looks like our CI is kaput, but I'll merge anyway.
